### PR TITLE
Transferexport

### PR DESCRIPTION
**Pull request by jernsting**
Added support for exporttree functions.

datalad / annex export works as expected, but leading dot is removed by dataverse (see #83).

### CODE REVIEW

**Comment by mih** [comment]
Looks wonderfully simple. Thx! Will have a closer look tomorrow.

File: datalad_dataverse/remote.py [hunk 1]
```diff
@@ -120,6 +134,9 @@ def transfer_retrieve(self, key, file):
         with open(file, "wb") as f:
             f.write(response.content)
 
+    def transferexport_retrieve(self, key, local_file, remote_file):
+        self.transfer_retrieve(remote_file, local_file)
```
Comment (bpoldrack) [inline datalad_dataverse/remote.py:138]: Minor remark on occasion of that line (but true for other calls to the non-export methods as well):
I think argument assignments can turn pretty confusing for someone having to occasionally maintain this code. I think I'd prefer using keywords to be explicit here - `self.transfer_retrieve(key=remote_file, file=local_file)`

Comment (jernsting) [reply datalad_dataverse/remote.py]: Done

**Comment by jadecci** [comment]
Added directory name validation (#17)

File: datalad_dataverse/remote.py [hunk 2]
```diff
@@ -6,14 +6,17 @@
...
     get_native_api,
 )
+import os
+import re
 
 from datalad_dataverse.utils import format_doi
 
 
-class DataverseRemote(SpecialRemote):
+class DataverseRemote(ExportRemote):
```
Comment (mih) [inline datalad_dataverse/remote.py:19]: This is now getting the class straight from AnnexRemote, and we are loosing `.message()` with that. Presently no other choice it seems. Just FTR

File: datalad_dataverse/remote.py [hunk 3]
```diff
@@ -77,18 +80,34 @@ def checkpresent(self, key):
...
+        if datafile is None:
+            datafile = Datafile()
+            datafile.set({'filename': key, 'label': key})
+        datafile.set({'pid': ds_pid})
+        
+        resp = self.api.upload_datafile(identifier=ds_pid, filename=local_file, json_str=datafile.json())
+        resp.raise_for_status()
+
+    def transferexport_store(self, key, local_file, remote_file):
+        if re.search(pattern='[^a-z0-9_\-.\\/\ ]', string=os.path.dirname(remote_file), flags=re.ASCII | re.IGNORECASE):
```
Comment (mih) [inline datalad_dataverse/remote.py:102]: Call to `os.path.dirname()` is duplicated below, better to do it once, and use a variable.

Comment (jadecci) [reply datalad_dataverse/remote.py]: Fixed now :)

File: datalad_dataverse/remote.py (hunk 2 above)
Comment (jernsting) [reply datalad_dataverse/remote.py:19]: We could inherit from both classes..? That would retain the `.message()` and the ExportRemote functionallity

Comment (jernsting) [inline datalad_dataverse/remote.py:19]:
```suggestion
from datalad.customremotes import SpecialRemote

class DataverseRemote(ExportRemote, SpecialRemote):
```

Comment (bpoldrack) [reply datalad_dataverse/remote.py:23]: Agree - inheriting from both seems right at a quick glance

**Review by bpoldrack — APPROVED** [review]
Looks good to me.
Thank you very much!

**Pull request merged**